src/target/msp432e4.c

```diff
@@ -0,0 +1,310 @@
+ */
+#define MSP432E4_FLASH_FCRIS (MSP432E4_FLASH_CTRL_BASE + 0x000cU)
+
+/*
+ * Flash Write Key
+ *  [31:16] - 0000 - reserved
+ *  [15:0] - xxxx - key
+ */
+#define MSP432E4_FLASH_FLPEKEY (MSP432E4_FLASH_CTRL_BASE + 0x003cU)
+
+/*
+ * Flash Access Address
+ *  [31:20] - 000 - reserved
+ *  [19:0] - xxxxx - operation aligned address
+ */
+#define MSP432E4_FLASH_ADDR (MSP432E4_FLASH_CTRL_BASE + 0x0000U)
+
+/* Flash Data */
+#define MSP432E4_FLASH_DATA (MSP432E4_FLASH_CTRL_BASE + 0x0004U)
+
+/*
+ * EEPROM Size
+ *  [31:16] - xxxx - # 16bit words
+ *  [15:0] - xxxx - # 32bit words
+ */
+#define MSP432E4_EEPROM_SIZE (MSP432E4_EEPROM_BASE + 0x0000U)
+
+#define MSP432E4_SRAM_BASE  0x20000000U
+#define MSP432E4_FLASH_BASE 0x00000000U
+
+/*
+ * XXX: The flash block size might be nice to use, but the part used in
+ * testing has a 16k sector which is too big w/ the current build setup.
+ */
+#define BUFFERED_WRITE_SIZE 0x100
+
+typedef struct msp432e4_flash {
+	target_flash_s target_flash;
+	uint16_t flash_key;
+} msp432e4_flash_s;
+
+static bool msp432e4_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len);
+static bool msp432e4_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len);
+static bool msp432e4_mass_erase(target_s *const target);
+
+static void msp432e4_add_flash(
+	target_s *const target, const uint32_t sector_size, const uint32_t base, const size_t length)
+{
+	msp432e4_flash_s *const flash = calloc(1, sizeof(*flash));
+	if (flash == NULL) {
+		DEBUG_WARN("calloc: failed in %s\n", __func__);
+		return;
+	}
+
+	target_flash_s *target_flash = &flash->target_flash;
+	target_flash->start = base;
+	target_flash->length = length;
+	target_flash->blocksize = sector_size;
+	target_flash->erase = msp432e4_flash_erase;
+	target_flash->write = msp432e4_flash_write;
+	target_flash->writesize = BUFFERED_WRITE_SIZE;
+	target_flash->erased = 0xff;
+	target_add_flash(target, target_flash);
+
+	/* If the boot config KEY bit is set, use the fixed key value, otherwise read out the configured key */
+	if (target_mem_read32(target, MSP432E4_SYS_CTRL_BOOTCFG) & MSP432E4_SYS_CTRL_BOOTCFG_KEY)
+		flash->flash_key = 0xa442U;
+	else
+		flash->flash_key = (uint16_t)target_mem_read32(target, MSP432E4_FLASH_FLPEKEY);
+}
+
+bool msp432e4_probe(target_s *const target)
+{
+	const uint32_t devid0 = target_mem_read32(target, MSP432E4_SYS_CTRL_DID0);
+	const uint32_t devid1 = target_mem_read32(target, MSP432E4_SYS_CTRL_DID1);
+	DEBUG_INFO("%s: devid %x:%x\n", __func__, devid0, devid1);
+
+	/* Does it look like an msp432e4 variant? */
+	if ((devid0 & MSP432E4_SYS_CTRL_DID0_MASK) != MSP432E4_SYS_CTRL_DID0_MSP432E4 ||
+		(devid1 & MSP432E4_SYS_CTRL_DID1_MASK) != MSP432E4_SYS_CTRL_DID1_MSP432E4)
+		return false;
+
+	DEBUG_INFO("%s: ver %x:%x part %x pin %x temp %x package %x\n", __func__, (devid0 >> 8) & 0xff,
+		(devid0 >> 0) & 0xff, (devid1 >> 16) & 0xff, (devid1 >> 13) & 0x7, (devid1 >> 5) & 0x7, (devid1 >> 3) & 0x3);
```

Comment:
Agreed on the offsets and masks. Rather than entirely removing this, though, it might be worth making it a DEBUG_TARGET() level thing, which would also allow the reported string to be a bit more meaningful. That won't then affect Flash usage on the main platforms but will make part identification nicer in logs from BMDA and Farpatch